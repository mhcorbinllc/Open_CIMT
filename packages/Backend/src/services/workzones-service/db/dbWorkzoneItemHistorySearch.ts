import {
  DynaMongoDB,
  findLikeSearchEngine,
} from "dyna-mongo-db";
import {
  IWorkzoneHistory,
  IDBWorkzoneHistory,
  convertIDBWorkzoneHistorytoIWorkzoneHistory,
} from "../interfaces/IWorkzoneHistory";
import {getWorkzonesItemsHistoryCollection} from "./getWorkzonesCollections";

export const dbWorkzoneItemHistorySearch = async (
  {
    dmdb,
    companyId,
    workzoneId,
    searchText,
    skip,
    limit,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    workzoneId: string;
    searchText?: string;
    skip: number;
    limit: number;
  },
): Promise<IWorkzoneHistory[]> => {
  const collection = await getWorkzonesItemsHistoryCollection(dmdb, companyId);

  const workzoneHistory: IDBWorkzoneHistory[] = await collection
    .find(
      {
        "workzone.workzoneId": workzoneId,
        ...findLikeSearchEngine('dbSearchText', searchText),
      },
      {
        skip,
        limit,
      },
    )
    .toArray();

  return workzoneHistory.map(convertIDBWorkzoneHistorytoIWorkzoneHistory);
};
