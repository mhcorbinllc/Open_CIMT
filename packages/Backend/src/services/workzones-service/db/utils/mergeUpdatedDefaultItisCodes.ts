import {Db} from "mongodb";
import {syncPromises} from "dyna-sync";

import {
  IWorkzonesAppSettings,
} from "../../interfaces/IWorkzonesAppSettings";
import {getDefaultAppItisCodes} from "../../interfaces/getDefaultAppItisCodes";

export const mergeUpdatedDefaultItisCodes = async (
  {
    db,
    collectionName,
  }: {
    db: Db;
    collectionName: string;
  },
): Promise<void> => {
  const collection = db.collection<IWorkzonesAppSettings>(collectionName);
  const appsSettings = await collection.find().toArray();
  const defaultAppItisCodes = getDefaultAppItisCodes();

  await syncPromises(
    ...appsSettings.map(appSettings => {

      const hasSameCodes =
        appSettings.itisCodes.map(v => v.code).sort()
          .join(',') ===
        defaultAppItisCodes.map(v => v.code).sort()
          .join(',');

      if (hasSameCodes) return () => Promise.resolve();

      const mergedItisCodes = defaultAppItisCodes
        .filter((v) => !appSettings.itisCodes.find(t =>
          t.code === v.code
          && t.label.toLowerCase() === v.label.toLowerCase(),
        ))
        .concat(appSettings.itisCodes);

      return async () => {
        await collection.updateOne(
          {deviceId: appSettings.deviceId},
          {$set: {itisCodes: mergedItisCodes}},
        );
      };
    }),
  );
};
