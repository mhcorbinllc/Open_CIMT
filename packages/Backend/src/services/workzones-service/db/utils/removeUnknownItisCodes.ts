import {IDBWorkzone} from "../../interfaces/IDBWorkzone";
import {DynaMongoDB} from "dyna-mongo-db";
import {dbWorkzonesAppSettingsLoad} from "../dbWorkzonesAppSettingsLoad";

export const removeUnknownItisCodes = async (
  {
    dmdb,
    companyId,
    dbWorkzones,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    dbWorkzones: IDBWorkzone[];
  },
): Promise<void> => {
  const settings = await dbWorkzonesAppSettingsLoad(dmdb, companyId);
  dbWorkzones
    .forEach(dbWorkzone => {
      dbWorkzone.itisCodes =
        dbWorkzone.itisCodes
          .filter(itisCode => settings.itisCodes.find(ic => itisCode === ic.code));
    });
};
