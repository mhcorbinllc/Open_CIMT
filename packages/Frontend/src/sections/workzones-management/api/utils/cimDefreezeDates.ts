import {IWorkzone} from "mhc-server";

export const cimDefreezeDates = (cim: IWorkzone): IWorkzone => {
  return {
    ...cim,
    start: new Date(cim.start),
    end: new Date(cim.end),
  };
};
