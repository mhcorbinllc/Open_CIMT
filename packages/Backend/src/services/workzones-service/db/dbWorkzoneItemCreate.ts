import {DynaMongoDB} from "dyna-mongo-db";
import {dynaError} from "dyna-error";
import {guid} from "dyna-guid";

import {IWorkzone} from "../interfaces/IWorkzone";
import {
  IDBWorkzone,
  convertIWorkzoneToIDBWorkzone,
  convertIDBWorkzoneToIWorkzone,
} from "../interfaces/IDBWorkzone";

import {getWorkzonesItemsCollection} from "./getWorkzonesCollections";
import {dbWorkzoneItemHistoryAdd} from "./dbWorkzoneItemHistoryAdd";
import {removeUnknownItisCodes} from "./utils/removeUnknownItisCodes";

export const dbWorkzoneItemCreate = async (
  {
    dmdb,
    companyId,
    userId,
    workzone,
    externalImport = false,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    userId: string;
    workzone: IWorkzone;
    externalImport?: boolean;
  },
): Promise<IWorkzone> => {
  const collection = await getWorkzonesItemsCollection(dmdb, companyId);
  const dbWorkzone: IDBWorkzone = convertIWorkzoneToIDBWorkzone(workzone);

  if (externalImport) {
    [
      'workzoneId',
      'createdAt',
      'updatedAt',
      'deletedAt',
    ].forEach(fieldName => {
      if (workzone[fieldName] === undefined) {
        throw dynaError({
          message: `dbWorkzoneItemCreate: [${fieldName}] is not defined`,
          status: 400,
          validationErrors: {[fieldName]: 'Not defined'},
        });
      }
    });
  }
  else {
    dbWorkzone.workzoneId = guid();
    const now = Date.now();
    dbWorkzone.createdAt = now;
    dbWorkzone.updatedAt = now;
    dbWorkzone.deletedAt = 0;
  }

  await collection.insertOne(dbWorkzone);
  await dbWorkzoneItemHistoryAdd({
    dmdb,
    companyId,
    userId,
    historyName: 'CIM creation',
    dbWorkzone: dbWorkzone,
  });

  await removeUnknownItisCodes({
    dmdb,
    companyId,
    dbWorkzones: [dbWorkzone],
  });

  return convertIDBWorkzoneToIWorkzone(dbWorkzone);
};
