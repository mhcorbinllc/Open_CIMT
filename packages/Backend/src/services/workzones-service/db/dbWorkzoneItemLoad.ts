import {IWorkzone} from "../interfaces/IWorkzone";
import {convertIDBWorkzoneToIWorkzone} from "../interfaces/IDBWorkzone";
import {DynaMongoDB} from "dyna-mongo-db";
import {getWorkzonesItemsCollection} from "./getWorkzonesCollections";
import {removeUnknownItisCodes} from "./utils/removeUnknownItisCodes";

export const dbWorkzoneItemLoad = async (
  {
    dmdb,
    companyId,
    workzoneId,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    workzoneId: string;
  },
): Promise<IWorkzone | null> => {
  const collection = await getWorkzonesItemsCollection(dmdb, companyId);

  const dbWorkzone = await collection.findOne({workzoneId});
  if (!dbWorkzone) return null;

  await removeUnknownItisCodes({
    dmdb,
    companyId,
    dbWorkzones: [dbWorkzone],
  });

  return convertIDBWorkzoneToIWorkzone(dbWorkzone);
};
