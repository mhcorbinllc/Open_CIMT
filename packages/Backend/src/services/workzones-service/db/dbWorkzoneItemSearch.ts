import {DynaMongoDB} from "dyna-mongo-db";

import {IWorkzone} from "../interfaces/IWorkzone";
import {
  IDBWorkzone,
  convertIDBWorkzoneToIWorkzone,
} from "../interfaces/IDBWorkzone";

import {getWorkzonesItemsCollection} from "./getWorkzonesCollections";
import {dbSearchCollection} from "../../../db/DBEntity/utils/dbSearchCollection";
import {removeUnknownItisCodes} from "./utils/removeUnknownItisCodes";

export const dbWorkzoneItemSearch = async (
  {
    dmdb,
    companyId,
    searchText,
    forEver = 'both',
    deleted,
    skip,
    limit,
    sortByFieldName,
    sortDirection,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    searchText?: string;
    forEver?: boolean | 'both';
    deleted?: boolean | 'both';
    skip: number;
    limit: number;
    sortByFieldName?: string;
    sortDirection?: 0 | 1 | -1;
  },
): Promise<IWorkzone[]> => {
  const collection = await getWorkzonesItemsCollection(dmdb, companyId);

  const dbWorkzones = await dbSearchCollection<IDBWorkzone>({
    collection,

    searchText,
    deleted,

    find: forEver === 'both'
      ? undefined
      : {forEver},

    skip,
    limit,

    sortByFieldName,
    sortDirection,
  });

  await removeUnknownItisCodes({
    dmdb,
    companyId,
    dbWorkzones,
  });

  return dbWorkzones.map(convertIDBWorkzoneToIWorkzone);
};
