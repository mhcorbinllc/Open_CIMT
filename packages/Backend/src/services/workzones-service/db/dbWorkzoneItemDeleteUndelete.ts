import {DynaMongoDB} from "dyna-mongo-db";

import {IWorkzone} from "../interfaces/IWorkzone";
import {getWorkzonesItemsCollection} from "./getWorkzonesCollections";
import {dbWorkzoneItemHistoryAdd} from "./dbWorkzoneItemHistoryAdd";
import {dbWorkzoneItemLoad} from "./dbWorkzoneItemLoad";
import {convertIWorkzoneToIDBWorkzone} from "../interfaces/IDBWorkzone";

export const dbWorkzoneItemDeleteUndelete = async (
  {
    dmdb,
    companyId,
    userId,
    operation,
    workzoneId,
    externalImport,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    userId: string;
    operation: 'delete' | 'undelete';
    workzoneId: string;
    externalImport?: {
      workzone: IWorkzone;
      newerOnly: boolean;
    };
  },
): Promise<{
  workzone: IWorkzone | null; // Null if not found
  applied: boolean;           // If the externalImportWorkzone is newer than server's version
}> => {
  const collection = await getWorkzonesItemsCollection(dmdb, companyId);
  const actionDate: number =
    externalImport
      ? externalImport.workzone.updatedAt
      : Date.now();

  const currentWorkzone = await dbWorkzoneItemLoad({
    dmdb,
    companyId,
    workzoneId,
  });

  if (!currentWorkzone) {
    return {
      workzone: null,
      applied: false,
    };
  }

  if (
    externalImport
    && externalImport.newerOnly
    && currentWorkzone.updatedAt > externalImport.workzone.updatedAt
  ) {
    return {
      workzone: currentWorkzone,
      applied: false,
    };
  }

  const updatedWorkzone: IWorkzone = {
    ...currentWorkzone,
    kapschId: operation === 'undelete' ? -1 : currentWorkzone.kapschId,
    updatedAt: actionDate,
    deletedAt: operation === 'delete' ? actionDate : 0,
  };

  await collection
    .updateOne(
      {workzoneId},
      {$set: updatedWorkzone},
    );

  const historyName =
      operation === 'delete'
        ? updatedWorkzone.kapschId > -1
          ? `CIM and broadcasted CIM kapschId ${updatedWorkzone.kapschId} deleted`
          : `CIM deleted`
        : 'CIM undeleted';

  await dbWorkzoneItemHistoryAdd({
    dmdb,
    companyId,
    userId,
    historyName,
    dbWorkzone: convertIWorkzoneToIDBWorkzone(updatedWorkzone),
  });

  return {
    workzone: updatedWorkzone,
    applied: true,
  };
};
