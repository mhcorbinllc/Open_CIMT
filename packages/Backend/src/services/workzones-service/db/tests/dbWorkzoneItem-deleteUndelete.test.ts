import {syncPromises} from "dyna-sync";
import {testDMDB} from "../../../../../tests/utils/testDMDB";

import {clearDynamicValuesForSnapshots} from "../../../../utils/clearDynamicValuesForSnapshots";

import {testWorkzones} from "./data/testWorkzones";

import {dbWorkzoneItemCreate} from "../dbWorkzoneItemCreate";
import {dbWorkzoneItemSearch} from "../dbWorkzoneItemSearch";
import {dbWorkzoneItemLoad} from "../dbWorkzoneItemLoad";
import {dbWorkzoneItemDeleteUndelete} from "../dbWorkzoneItemDeleteUndelete";
import {dbWorkzoneItemHistorySearch} from "../dbWorkzoneItemHistorySearch";

describe('CIM Items - Create Update', () => {
  const companyId = 'test-company--test-480246';
  const userId = 'test-user-id-23242';

  const c = (data: any): any => clearDynamicValuesForSnapshots(data, 'workzoneId'.split(','));

  const dmdb = testDMDB;
  const clearDb = async () => {
    try {
      await dmdb.dropCollection([companyId, 'workzones-items'].join('---'));
      await dmdb.dropCollection([companyId, 'workzones-items-history'].join('---'));
    }
    catch (e) {
      console.error('clearDb error:', e);
    }
  };

  beforeAll(async () => {
    await clearDb();
  });
  afterAll(async () => {
    await clearDb();
    await dmdb.disconnect();
  });

  test('Create Update Check CIM item the history', async () => {
    await syncPromises(
      ...testWorkzones
        .map(workzone => () => dbWorkzoneItemCreate({
          dmdb,
          companyId,
          userId,
          workzone,
        })),
    );

    const loadWorkzones = await dbWorkzoneItemSearch({
      dmdb,
      companyId,
      skip: 0,
      limit: 100,
    });

    expect(loadWorkzones.length).toBe(8);

    const searchWorkzoneWoods204 = await dbWorkzoneItemSearch({
      dmdb,
      companyId,
      searchText: "Woods 204",
      skip: 0,
      limit: 100,
    });

    expect(searchWorkzoneWoods204.length).toBe(1);

    const workzoneWoods204 = searchWorkzoneWoods204[0];

    await dbWorkzoneItemDeleteUndelete({
      dmdb,
      companyId,
      userId,
      workzoneId: workzoneWoods204.workzoneId,
      operation: 'delete',
    });

    const searchForWorkzoneWoods204 = await dbWorkzoneItemSearch({
      dmdb,
      companyId,
      searchText: "Woods 204",
      skip: 0,
      limit: 100,
    });

    expect(searchForWorkzoneWoods204.length).toBe(0);

    const loadDeletedWoods204 = await dbWorkzoneItemLoad({
      dmdb,
      companyId,
      workzoneId: workzoneWoods204.workzoneId,
    });

    expect(loadDeletedWoods204).not.toBe(null);
    expect(c(loadDeletedWoods204)).toMatchSnapshot('Deleted CIM');

    const searchForWorkzoneWoods204WhereIsDeleted = await dbWorkzoneItemSearch({
      dmdb,
      companyId,
      searchText: "Woods 204",
      skip: 0,
      limit: 100,
    });

    expect(searchForWorkzoneWoods204WhereIsDeleted.length).toBe(0);

    const searchForWorkzoneWoods204AsDeleted = await dbWorkzoneItemSearch({
      dmdb,
      companyId,
      searchText: "Woods 204",
      deleted: true,
      skip: 0,
      limit: 100,
    });

    expect(searchForWorkzoneWoods204AsDeleted.length).toBe(1);

    await dbWorkzoneItemDeleteUndelete({
      dmdb,
      companyId,
      userId,
      workzoneId: workzoneWoods204.workzoneId,
      operation: 'undelete',
    });

    const searchForWorkzoneTheUndeletedWoods204 = await dbWorkzoneItemSearch({
      dmdb,
      companyId,
      searchText: "Woods 204",
      deleted: false,
      skip: 0,
      limit: 100,
    });

    expect(searchForWorkzoneTheUndeletedWoods204.length).toBe(1);
    expect(c(searchForWorkzoneTheUndeletedWoods204[0])).toMatchSnapshot('The undeleted CIM');

    const woods204History = await dbWorkzoneItemHistorySearch({
      dmdb,
      companyId,
      workzoneId: workzoneWoods204.workzoneId,
      skip: 0,
      limit: 100,
    });

    expect(woods204History.length).toBe(3);
    expect(c(woods204History)).toMatchSnapshot('History');
  });
});
