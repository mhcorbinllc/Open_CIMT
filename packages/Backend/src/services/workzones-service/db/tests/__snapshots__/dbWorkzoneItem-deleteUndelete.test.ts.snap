// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CIM Items - Create Update Create Update Check CIM item the history: Deleted CIM 1`] = `
Object {
  "active": true,
  "broadcast": Object {
    "selectionByRadius": Object {
      "radiusInMeters": 0,
    },
    "selectionByRegion": Object {
      "region": Object {
        "points": Array [
          Object {
            "lat": -0.232,
            "lng": 1,
          },
        ],
      },
    },
    "selectionMode": "REGION",
  },
  "closedLane": "LEFT",
  "closedLaneWidthInMeters": 0,
  "closedShoulder": "RIGHT",
  "closureSelectionMode": "CUSTOM",
  "closureType": "NONE",
  "createdAt": "---NUMBER>0---",
  "deletedAt": "---NUMBER>0---",
  "end": "2021-04-15T22:00:00.000Z",
  "forEver": false,
  "heading": Array [],
  "itisCodes": Array [],
  "kapschId": 24058204,
  "name": "Woods 204 - 621 Hutton",
  "notes": "Some additional info about this.
Road defect",
  "path": Object {
    "points": Array [],
  },
  "point": Object {
    "lat": -1,
    "lng": -1,
  },
  "speedLimitInMiles": 10,
  "start": "2021-04-01T20:00:00.000Z",
  "updatedAt": "---NUMBER>0---",
  "workersPresent": true,
  "workzoneId": "---STRING_NON_EMPTY---",
}
`;

exports[`CIM Items - Create Update Create Update Check CIM item the history: History 1`] = `
Array [
  Object {
    "date": "---NUMBER>0---",
    "name": "CIM creation",
    "userId": "test-user-id-23242",
    "workzone": Object {
      "active": true,
      "broadcast": Object {
        "selectionByRadius": Object {
          "radiusInMeters": 0,
        },
        "selectionByRegion": Object {
          "region": Object {
            "points": Array [
              Object {
                "lat": -0.232,
                "lng": 1,
              },
            ],
          },
        },
        "selectionMode": "REGION",
      },
      "closedLane": "LEFT",
      "closedLaneWidthInMeters": 0,
      "closedShoulder": "RIGHT",
      "closureSelectionMode": "CUSTOM",
      "closureType": "NONE",
      "createdAt": "---NUMBER>0---",
      "deletedAt": "---NUMBER=0---",
      "end": "2021-04-15T22:00:00.000Z",
      "forEver": false,
      "heading": Array [],
      "itisCodes": Array [],
      "kapschId": 24058204,
      "name": "Woods 204 - 621 Hutton",
      "notes": "Some additional info about this.
Road defect",
      "path": Object {
        "points": Array [],
      },
      "point": Object {
        "lat": -1,
        "lng": -1,
      },
      "speedLimitInMiles": 10,
      "start": "2021-04-01T20:00:00.000Z",
      "updatedAt": "---NUMBER>0---",
      "workersPresent": true,
      "workzoneId": "---STRING_NON_EMPTY---",
    },
  },
  Object {
    "date": "---NUMBER>0---",
    "name": "CIM and broadcasted CIM kapschId 24058204 deleted",
    "userId": "test-user-id-23242",
    "workzone": Object {
      "active": true,
      "broadcast": Object {
        "selectionByRadius": Object {
          "radiusInMeters": 0,
        },
        "selectionByRegion": Object {
          "region": Object {
            "points": Array [
              Object {
                "lat": -0.232,
                "lng": 1,
              },
            ],
          },
        },
        "selectionMode": "REGION",
      },
      "closedLane": "LEFT",
      "closedLaneWidthInMeters": 0,
      "closedShoulder": "RIGHT",
      "closureSelectionMode": "CUSTOM",
      "closureType": "NONE",
      "createdAt": "---NUMBER>0---",
      "deletedAt": "---NUMBER>0---",
      "end": "2021-04-15T22:00:00.000Z",
      "forEver": false,
      "heading": Array [],
      "itisCodes": Array [],
      "kapschId": 24058204,
      "name": "Woods 204 - 621 Hutton",
      "notes": "Some additional info about this.
Road defect",
      "path": Object {
        "points": Array [],
      },
      "point": Object {
        "lat": -1,
        "lng": -1,
      },
      "speedLimitInMiles": 10,
      "start": "2021-04-01T20:00:00.000Z",
      "updatedAt": "---NUMBER>0---",
      "workersPresent": true,
      "workzoneId": "---STRING_NON_EMPTY---",
    },
  },
  Object {
    "date": "---NUMBER>0---",
    "name": "CIM undeleted",
    "userId": "test-user-id-23242",
    "workzone": Object {
      "active": true,
      "broadcast": Object {
        "selectionByRadius": Object {
          "radiusInMeters": 0,
        },
        "selectionByRegion": Object {
          "region": Object {
            "points": Array [
              Object {
                "lat": -0.232,
                "lng": 1,
              },
            ],
          },
        },
        "selectionMode": "REGION",
      },
      "closedLane": "LEFT",
      "closedLaneWidthInMeters": 0,
      "closedShoulder": "RIGHT",
      "closureSelectionMode": "CUSTOM",
      "closureType": "NONE",
      "createdAt": "---NUMBER>0---",
      "deletedAt": "---NUMBER=0---",
      "end": "2021-04-15T22:00:00.000Z",
      "forEver": false,
      "heading": Array [],
      "itisCodes": Array [],
      "kapschId": -1,
      "name": "Woods 204 - 621 Hutton",
      "notes": "Some additional info about this.
Road defect",
      "path": Object {
        "points": Array [],
      },
      "point": Object {
        "lat": -1,
        "lng": -1,
      },
      "speedLimitInMiles": 10,
      "start": "2021-04-01T20:00:00.000Z",
      "updatedAt": "---NUMBER>0---",
      "workersPresent": true,
      "workzoneId": "---STRING_NON_EMPTY---",
    },
  },
]
`;

exports[`CIM Items - Create Update Create Update Check CIM item the history: The undeleted CIM 1`] = `
Object {
  "active": true,
  "broadcast": Object {
    "selectionByRadius": Object {
      "radiusInMeters": 0,
    },
    "selectionByRegion": Object {
      "region": Object {
        "points": Array [
          Object {
            "lat": -0.232,
            "lng": 1,
          },
        ],
      },
    },
    "selectionMode": "REGION",
  },
  "closedLane": "LEFT",
  "closedLaneWidthInMeters": 0,
  "closedShoulder": "RIGHT",
  "closureSelectionMode": "CUSTOM",
  "closureType": "NONE",
  "createdAt": "---NUMBER>0---",
  "deletedAt": "---NUMBER=0---",
  "end": "2021-04-15T22:00:00.000Z",
  "forEver": false,
  "heading": Array [],
  "itisCodes": Array [],
  "kapschId": -1,
  "name": "Woods 204 - 621 Hutton",
  "notes": "Some additional info about this.
Road defect",
  "path": Object {
    "points": Array [],
  },
  "point": Object {
    "lat": -1,
    "lng": -1,
  },
  "speedLimitInMiles": 10,
  "start": "2021-04-01T20:00:00.000Z",
  "updatedAt": "---NUMBER>0---",
  "workersPresent": true,
  "workzoneId": "---STRING_NON_EMPTY---",
}
`;
